### PR TITLE
Collection of small changes

### PR DESCRIPTION
**Pull request by JackTemaki**
I wrote some small commits that I did not want to push directly into the master branch without review. I thought it is easier to bundle it in a single pull request.

### CODE REVIEW

File: TFNetwork.py [hunk 1]
```diff
@@ -985,6 +985,9 @@ def inv_reduce_sum(x, name):
           d["cost:%s" % loss_name] = reduce_sum(loss.get_loss_value_for_fetch(), name="cost:%s" % loss_name)
         if loss.get_error_value() is not None:
           d["error:%s" % loss_name] = reduce_sum(loss.get_error_value(), name="error:%s" % loss_name)
+        if config.bool("debug_display_loss_per_layer", False):
+          if loss.get_loss_value_for_objective() is not None:
+            d["loss:%s" % loss_name] = reduce_sum(loss.get_loss_value_for_objective(), name="loss:%s" % loss_name)
```
Comment (albertz) [inline TFNetwork.py]: The key name is confusing. It's already confusing that we have sort of "cost" == "loss", but this makes it more confusing.
Better name it `"cost_objective:..."` or so.
But anyway, is this really needed? I think adding it via `tf.summary.scalar` is probably better, right?
Note also that the difference is not necessarily that one is normalized and the other is not.

File: TFEngine.py [hunk 2]
```diff
@@ -276,7 +276,8 @@ def _collect_eval_info(self, fetches_results):
     """
     # See see self._get_fetches_dict() for the keys.
     # keys are e.g. "cost:output", "error:output" or "loss".
-    keys = [k for k in fetches_results.keys() if k.startswith("cost:") or k.startswith("error:") or k == "loss"]
+    keys = [k for k in fetches_results.keys() if k.startswith("cost:") or k.startswith("error:") or
+            k.startswith("loss:") or k == "loss"]
```
Comment (albertz) [inline TFEngine.py]: The formatting is confusing. `if ...` should be on the same line with all the conditions (all the `or`). If the line becomes too long, make the newline before the `if`.

File: TFNetworkRecLayer.py [hunk 3]
```diff
@@ -4670,10 +4670,13 @@ def _weights_time_axis_to_reduce(cls, weights, base):
     # SoftmaxOverSpatialLayer by default uses time_dim_axis.
     # We also should maybe check that this matches the base time dim axis.
     dyn_axes = weights.get_dynamic_axes()
+
```
Comment (albertz) [inline TFNetworkRecLayer.py]: Why this extra empty line?

File: TFNetworkRecLayer.py [hunk 4]
```diff
@@ -4670,10 +4670,13 @@ def _weights_time_axis_to_reduce(cls, weights, base):
...
     # for static time-dim
-    if weights.time_dim_axis not in dyn_axes:
+    if not dyn_axes or weights.time_dim_axis not in dyn_axes:
+      assert weights.time_dim_axis is not None and weights.time_dim_axis is not NotSpecified
       return weights.time_dim_axis
     assert dyn_axes, "no dynamic axes in %r" % weights
+
```
Comment (albertz) [inline TFNetworkRecLayer.py]: Why this extra empty line?

File: TFNetworkRecLayer.py [hunk 5]
```diff
@@ -4670,10 +4670,13 @@ def _weights_time_axis_to_reduce(cls, weights, base):
     # SoftmaxOverSpatialLayer by default uses time_dim_axis.
     # We also should maybe check that this matches the base time dim axis.
     dyn_axes = weights.get_dynamic_axes()
+
     # for static time-dim
-    if weights.time_dim_axis not in dyn_axes:
+    if not dyn_axes or weights.time_dim_axis not in dyn_axes:
```
Comment (albertz) [inline TFNetworkRecLayer.py]: This change (adding `not dyn_axes`) does not change anything. Why do you change it then? It's confusing now, because `not dyn_axes` is always False if the other condition is False, and if the other condition is True, it does not matter.

File: TFNetworkRecLayer.py [hunk 6]
```diff
@@ -4670,10 +4670,13 @@ def _weights_time_axis_to_reduce(cls, weights, base):
...
+      assert weights.time_dim_axis is not None and weights.time_dim_axis is not NotSpecified
```
Comment (albertz) [inline TFNetworkRecLayer.py]: It never can be `NotSpecified`. That check does not make sense.
Also, I don't quite understand how this fixes a problem.

Comment (JackTemaki) [reply TFNetworkRecLayer.py]: This seems to be a merge error, as the check for static time axis did not exist when I made the change a long time ago (but never did a commit for it before), so the commit can be removed

Comment (albertz) [reply TFNetworkRecLayer.py]: Ok, then remove it, and update the PR.

Comment (albertz) [reply TFNetworkRecLayer.py]: And resolve this when done.

File: TFNetwork.py (hunk 1 above)
Comment (JackTemaki) [reply TFNetwork.py]: I used this to be able to compare the actual loss of two different objectives and tune my loss scaling accordingly. The "cost" values do not include the loss scaling and seemed to be normalized with respect to the sequence length. I did not see another way to solve my issue.

Comment (albertz) [reply TFNetwork.py]: This does not address any of my points?

* The key name is confusing. It's already confusing that we have sort of "cost" == "loss", but this makes it more confusing.
Better name it "cost_objective:..." or so.
* Is `tf.summary.scalar` maybe better? (That would have the advantage that the code is much simpler.)
* Note also that the difference is not necessarily that one is normalized and the other is not.

Comment (JackTemaki) [reply TFNetwork.py]: I refered to "But anyway, is this really needed?". Besides of how to use tf.summary.scalar (the tensorflow page did not help), I understand your points.

Comment (albertz) [reply TFNetwork.py]: Do you know TensorBoard? See TF documentation about that. Putting stuff on stdout is somewhat limited, and I think for more extra info, TensorBoard (via `tf.summary`) should always be used. That's what we used already a lot (e.g. stuff like gradient norms, or basically anything).

Comment (JackTemaki) [reply TFNetwork.py]: Yes, I used Tensorboard once but it did not show more than the logs/learning-rate file itself for me, so I never used it again (are there flags for enabling more values?). But if you want it do be added only there it is ok, and I guess the option can be removed. The question is then how to distinguish the losses? "score" (in the learning rates file), "cost" (log) and "loss" (tensorboard) are all the same, so how should I call the actual loss? "objective_loss"? Also, they would be equal when `loss.use_normalized_loss` is `True`, but that should be ok.

Comment (albertz) [reply TFNetwork.py]: TensorBoard should always show all losses.
See all `tf.summary.scalar` usages.
You will find (among other things) this:
```
tf.summary.scalar("loss_%s" % name, self._loss_value * self._norm_factor)
```
If you have your debug option enabled, you could add this:
```
tf.summary.scalar("loss_unnormalized_%s" % name, self._loss_value)
```

Comment (JackTemaki) [reply TFNetwork.py]: Yes, I already checked the Tensorboard again, and also the function `_tf_summary` of the `LossHolder`. My question was how you can enable more than the all losses that are already part of the logs. (e.g, the gradient norms you mentioned)

Also unnormalized might be the wrong term, as (if I understand correctly) the actual loss can be normalized. Or do you want to display normalized and un-normalized, and let the user have to know which is used for gradient computation and which is not?

Comment (albertz) [reply TFNetwork.py]: Just see all usages of `tf.summary.scalar`. E.g. there is `debug_grad_summaries`.
`_loss_value` at that point is not normalized. In any case, the only difference is that you have factor `self._norm_factor` in one case and not in the other case, so I think it is a reasonable name (or `self._loss_value * self._norm_factor` would be wrong then, right?).

Comment (JackTemaki) [reply TFNetwork.py]: I am just saying that my initial code had a slightly different behaviour. But yes, I can add it as always normalized and un-normalized. I will also add `debug_grad_summaries` to the docs.

**Comment by JackTemaki** [comment]
what do you think about this? I now added two flags, one for unnormalized losses and one for the actual objective loss (including the scaling, and normalized only if activated).

File: TFNetwork.py [hunk 7]
```diff
@@ -2071,6 +2071,10 @@ def _tf_summary(self):
         tf.summary.scalar("loss_%s" % name, self._loss_value * self._norm_factor)
         if self._network.get_config().bool("calculate_exp_loss", False):
           tf.summary.scalar("exp_loss_%s" % name, tf.exp(self._loss_value * self._norm_factor))
+        if self._network.get_config().bool("debug_unnormalized_loss_summaries", False):
+          tf.summary.scalar("unnormalized_loss_%s" % name, self._loss_value)
+        if self._network.get_config().bool("debug_objective_loss_summaries", False):
+          tf.summary.scalar("objective_loss_%s" % name, self._loss_value_for_objective)
```
Comment (albertz) [inline TFNetwork.py]: Did you test this? Do you actually use this? I think this does not work, as `_loss_value_for_objective` is not set at this point. I would just remove it, if you don't use it. Don't just add random features.
See where `_tf_summary` is called. The logic currently makes sure that `_loss_value` at that point is actually exactly the loss used for the objective. I think this option is confusing.

Comment (albertz) [reply TFNetwork.py]: Ah, I'm just seeing, you moved the `_tf_summary` call.
But anyway, I'm not really sure why you would need `objective_loss_...`. How is this useful? You either have the information in `loss_...` or `unnormalized_loss_...`.

File: TFEngine.py (hunk 2 above)
Comment (albertz) [reply TFEngine.py]: Please resolve this thread if it is resolved (or outdated).

File: TFNetworkRecLayer.py (hunk 6 above)
Comment (albertz) [reply TFNetworkRecLayer.py]: I guess this is resolved?

File: TFNetwork.py (hunk 7 above)
Comment (JackTemaki) [reply TFNetwork.py]: Normalized and unnormalized do not include the loss scaling. I use pretraining to modify the loss scaling over time, and i want to now the actual relation of the losses during the training. And for me it seemed only `_loss_value_for_objective` contains the exact loss for gradient computation in all cases.

Comment (albertz) [reply TFNetwork.py]: Ah I see, ok.
